File: 337.打家劫舍-iii.java
import java.util.Map;
import java.util.HashMap;
/*
 * @lc app=leetcode.cn id=337 lang=java
 *
 * [337] 打家劫舍 III
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private Map<TreeNode, Integer> memo;

    public int rob(TreeNode root) {
        memo = new HashMap<>();
        return helper(root);
    }

    private int helper(TreeNode root) {
        if (root == null)  return 0;
        if (memo.containsKey(root))  return memo.get(root);
        int case_rob = root.val
            + (root.left == null ? 0 : helper(root.left.left) + helper(root.left.right))
            + (root.right == null ? 0 : helper(root.right.left) + helper(root.right.right));
        int case_not_rob = helper(root.left) + helper(root.right);
        int ans = Math.max(case_rob, case_not_rob);
        memo.put(root, ans);
        return ans;
    }
}
// @lc code=end

File: 337.打家劫舍-iii.py
#
# @lc app=leetcode.cn id=337 lang=python3
#
# [337] 打家劫舍 III
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rob(self, root: TreeNode) -> int:
        def robNode(root):
            if not root:
                return 0, 0
            left_res = robNode(root.left)
            right_res = robNode(root.right)
            # If not rob current node, result is maximum sum of children results (regardless of rob or not)
            res_not_rob = max(left_res[0], left_res[1]) + max(right_res[0], right_res[1])
            # If rob current node, result is sum of current node and children `not_rob` results
            res_rob = left_res[0] + right_res[0] + root.val
            return res_not_rob, res_rob

        return max(robNode(root))
        
# @lc code=end

